src/integrationTest/java/com/mongodb/hibernate/junit/MongoExtension.java

```diff
@@ -0,0 +1,101 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.junit;
+
+import static org.junit.platform.commons.support.AnnotationSupport.findAnnotatedFields;
+import static org.junit.platform.commons.util.ReflectionUtils.isStatic;
+
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.hibernate.internal.cfg.MongoConfigurationBuilder;
+import java.lang.reflect.Field;
+import java.util.Map;
+import java.util.function.Predicate;
+import org.bson.BsonDocument;
+import org.hibernate.cfg.Configuration;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestInstanceFactoryContext;
+import org.junit.jupiter.api.extension.TestInstancePreConstructCallback;
+import org.junit.jupiter.api.extension.TestInstancePreDestroyCallback;
+
+public final class MongoExtension
+        implements TestInstancePreConstructCallback,
+                TestInstancePreDestroyCallback,
+                BeforeAllCallback,
+                BeforeEachCallback {
+
+    private ExtensionContext outermostContext;
+
+    private MongoClient mongoClient;
+    private MongoDatabase mongoDatabase;
+
+    @Override
+    public void preConstructTestInstance(TestInstanceFactoryContext factoryContext, ExtensionContext context) {
+        if (factoryContext.getOuterInstance().isEmpty()) {
+            outermostContext = context;
+
+            @SuppressWarnings("unchecked")
+            Map<String, Object> hibernateProperties =
+                    (Map<String, Object>) (Map<?, Object>) new Configuration().getProperties();
+
+            var mongoConfig = new MongoConfigurationBuilder(hibernateProperties).build();
+            mongoClient = MongoClients.create(mongoConfig.mongoClientSettings());
+            mongoDatabase = mongoClient.getDatabase(mongoConfig.databaseName());
+        }
+    }
+
+    @Override
+    public void preDestroyTestInstance(ExtensionContext context) {
+        if (context.equals(outermostContext)) {
+            if (mongoClient != null) {
+                mongoClient.close();
+            }
+        }
+    }
+
+    @Override
+    public void beforeAll(ExtensionContext context) {
+        if (context.getTestInstance().orElse(null) instanceof MongoDatabaseAware mongoDatabaseAware) {
+            mongoDatabaseAware.injectMongoDatabase(mongoDatabase);
+        }
+
+        Predicate<Field> predicate = field -> MongoCollection.class.isAssignableFrom(field.getType());
+        findAnnotatedFields(context.getRequiredTestClass(), InjectMongoCollection.class, predicate)
```

Comment:
The predicate you suggest hides mistakes. If we declare

```java
@InjectMongoCollection("books")
String collection;
```

Then with the current predicate the field stays `null`, but the test run does not fail. It is better for the test run to fail, which we can achieve by not filtering out felds:

```suggestion
        findAnnotatedFields(context.getRequiredTestClass(), InjectMongoCollection.class, field -> true)
```